### ISSUE TITLE
LLVM builder for ARM64

### ISSUE THREAD

**Issue by jjones-cavium** [WorksOnArm/cluster#24]
Joel Jones, joel.jones@cavium.com, Cavium, Team Lead Server Toolchain 
Stefan Teleman, stefan.teleman@cavium.com, Cavium, Toolchain Engineer

LLVM Builder for ARM64
Currently, there are only a few native builders in the LLVM buildbot farm for ARM64, which means that execution tests aren't run frequently. See [http://lab.llvm.org:8011](url). There are also no pre-built binaries for top-of-trunk, only for releases. The instructions for how to add a builder to the builtbot farm is at: http://llvm.org/docs/HowToAddABuilder.html.

Anyone who uses LLVM natively would benefit from having more frequent execution tests and pre-built top-of-trunk binaries.

The LLVM project is 100% open source and is located at http://llvm.org/svn/llvm-project/llvm/trunk

Given the rapidity of changes to LLVM, a fully dedicated machine with network access is needed.
I don't anticipate needing much support as long as we have shell access to the machine.

I have been a contributor to the LLVM project and GCC.

**Comment by vielmetti** [comment]
Approved! Looking forward to getting this going.

Please create a Packet account if you don't already have one, and then create a project. I'll need the UUID of that project (send it in email) in order to mark it as "sponsored". You'll want to have everyone on the team join the project so that all of their SSH keys are in the system before you create a new server. One person will have a "billing" relationship, but the project itself will be paid for by Works on Arm.

**Comment by jjones-cavium** [comment]
Ed:

I’ve created the project “LLVM builder for ARM64” with the Project-ID of  81a516a2-d336-4baa-97a3-0c9e4ebeb8fc

Joel Jones

On Nov 21, 2017, at 1:13 PM, Edward Vielmetti <notifications@github.com<mailto:notifications@github.com>> wrote:


Approved! Looking forward to getting this going.

Please create a Packet account if you don't already have one, and then create a project. I'll need the UUID of that project (send it in email) in order to mark it as "sponsored". You'll want to have everyone on the team join the project so that all of their SSH keys are in the system before you create a new server. One person will have a "billing" relationship, but the project itself will be paid for by Works on Arm.

**Comment by jjones-cavium** [comment]
I've created an account, created a project, invited Stefan to join, and emailed (I think) with the Project ID.

**Comment by vielmetti** [comment]
Should be all set to create your first system - give it a go!

**Comment by vielmetti** [comment]
@jjones-cavium , stefan -

I think that #42 is a dup of this request, and I'd like help sorting out how to best satisfy these two sets of requests in a coordinated way to avoid duplication of effort.

thanks

**Comment by jjones-cavium** [comment]
Cavium machines don’t support AArch32, so any LLVM builder wouldn’t be able to do run testing.

Joel

On Feb 8, 2018, at 10:55 AM, Edward Vielmetti <notifications@github.com<mailto:notifications@github.com>> wrote:


@jjones-cavium<https://github.com/jjones-cavium> , stefan -

I think that #42<https://github.com/WorksOnArm/cluster/issues/42> is a dup of this request, and I'd like help sorting out how to best satisfy these two sets of requests in a coordinated way to avoid duplication of effort.

thanks

**Comment by maxim-kuvyrkov** [comment]
Hi @jjones-cavium ,

TL;DR I think Linaro should continue to maintain aarch64 and aarch32 LLVM buildbots.  I have been working for the last few months to migrate bots to high-core-count machines from Cavium and HiSilicon.  I'm fine with hosting aarch64 bots on either Cavium or HiSilicon hardware, but we need HiSilicon hardware for aarch32 bots.

We at Linaro have been maintaining LLVM aarch64 and aarch32 bots for years, and it is not an easy task.  I have been experimenting with running LLVM buildbots on both Cavium ThX1 and HiSilicon Taishin machines since November, and it took me until now to get configuration of LLVM buildbot master, and buildbot slaves stable.
A big problem was with RAM usage in high-parallelism builds.  While most of the time a single 64-thread build requires about 32GB of RAM, there are peaks when it requires over 256GB RAM, which exceeds normal configuration of Packet machines.  I've managed to solve this by adjusting ninja build system to throttle builds on high RAM usage.
Over the last few months we have moving LLVM bots to docker containers so that we can host several bots on a single powerful machine.  This is now complete and I'm ready to deploy bots to Packet.net infrastructure.

**Comment by vielmetti** [comment]
Looping @leandron into this conversation because he also has an LLVM bot going, and has been looking to minimize build times.

**Comment by maxim-kuvyrkov** [comment]
@jjones-cavium , @leandron , one of the goal of llvm buildbot migration to "big" machines is that we will able to scale number of bots easily.  At the moment I'm running 30 bots (2 full sets of both aarch64 and aarch32 bots) on a single HiSilicon machine and they are chugging along at ~300 load average for 64-core machine.  Once the first builds finish and ccache is populated the builds will be much faster.

I'm happy to add more bot configurations that you are interested in.  E.g., like in here: https://reviews.llvm.org/D43119 .

**Comment by vielmetti** [comment]
Reopening this discussion to coordinate.

**Comment by vielmetti** [comment]
Moving this along - @maxim-kuvyrkov @jjones-cavium @leandron if you have any updates on buildbot status I'd appreciate it.

I've also updated https://github.com/WorksOnArm/cluster/blob/master/ACKNOWLEDGEMENT.md to make suggestions for acknowledging this contribution - we are of course happy to spread the word.

**Comment by leandron** [comment]
We've started testing the build servers on the upstream yesterday.

http://lab.llvm.org:8011/builders/clang-aarch64-linux-build-cache
http://lab.llvm.org:8011/builders/clang-armv7-linux-build-cache

We are in contact with @maxim-kuvyrkov regarding some code we want to contribute to the Docker containers being used to host the builedbot builders on Packet machines. This is an ongoing task.

**Comment by vielmetti** [comment]
Since the build servers are working, I'll call this project "complete" from the setup point of view so that stalebot doesn't bother us again. If something comes up we can reopen!
